src/circuit/hashmap.rs

```diff
@@ -0,0 +1,152 @@
+use bellman::pairing::{
+    Engine,
+};
+
+use bellman::{
+    SynthesisError,
+    ConstraintSystem
+};
+
+use super::{
+    Assignment
+};
+
+use super::num::AllocatedNum;
+
+use super::boolean::{
+    AllocatedBit,
+    Boolean
+};
+
+use super::permutation::{
+    PermutationElement,
+    SortablePermutationElement
+};
+
+use std::cmp::Ordering;
+
+use bellman::pairing::ff::PrimeField;
+
+impl<E: Engine> PermutationElement<E> for Vec<AllocatedNum<E>>
```

Comment:
This is a trivia conditional swap over vector and it’s already implemented in a circuit. In any case it should take Option<bool> anyway. It is not a permutation network